File: direct-mode/clients/flutter/lib/avatar_view_model.dart
import 'dart:async';
import 'dart:typed_data';

import 'package:flutter/material.dart';
import 'package:record/record.dart';
import 'package:spatius_avatarkit/spatius_avatarkit.dart' as ak;

import 'realtime_client.dart';

enum ToastKind { error, warning }

class ToastMessage {
  const ToastMessage(this.text, {this.kind = ToastKind.error});

  final String text;
  final ToastKind kind;
}

class AvatarViewModel extends ChangeNotifier {
  // --- Public state ---
  String connectionState = 'disconnected';
  String conversationState = 'idle';
  String? errorMessage;
  /// onFrameRateInfo — null until the monitor has reported once.
  int? fps;
  ak.Avatar? avatar;

  /// Set by the page so failures and blocked actions surface in the UI
  /// instead of only reaching [errorMessage].
  void Function(ToastMessage)? onToast;

  // --- Conversation ---
  /// Whether the agent socket is being opened, and whether it is ready to be spoken
  /// to. Audio pushed before ready is dropped, which presents as a microphone that
  /// records and is never answered.
  bool agentConnecting = false;
  bool agentReady = false;
  bool micActive = false;

  /// What has been said so far, as (role, text).
  List<(String, String)> transcript = [];

  /// Where the agent socket lives, as the server reported it. The language, the
  /// models and the voice are the server's `.env` — all three are fixed when it
  /// builds the agent session, so none of them is a client setting.
  String realtimeUrl = '';

  // --- Private ---
  ak.AvatarController? _controller;

  RealtimeClient? _realtime;

  final AudioRecorder _recorder = AudioRecorder();
  StreamSubscription<Uint8List>? _micSubscription;


  // --- Controller ---

  void setAvatarController(ak.AvatarController controller) {
    _controller = controller;

    // Off by default and free while off; the status bar is what asks for it.
    controller.setFrameRateMonitorEnabled(true);
    controller.onFrameRateInfo = (info) {
      // displayFps, not productionFps: what reached the screen, which is what a
      // reader comparing devices is actually asking about.
      fps = info.displayFps.round();
      notifyListeners();
    };

    controller.onConnectionState = (state, errorMsg) {
      connectionState = state.name;
      notifyListeners();
    };

    controller.onConversationState = (state) {
      conversationState = state.name;
      notifyListeners();
    };

    controller.onError = (error) {
      errorMessage = error.name;
      onToast?.call(ToastMessage(error.name));
      notifyListeners();
    };
  }

  // --- Lifecycle ---

  void start() => _controller?.start();

  void pause() => _controller?.pause();

  void resume() => _controller?.resume();

  void interrupt() {
    _controller?.interrupt();
  }

  // --- Conversation ---

  /// Open the agent socket, once per session.
  ///
  /// Opening the microphone is what brings it up: an agent costs a model session,
  /// so it is not started before there is something to say to it.
  Future<void> _ensureAgent() async {
    if (agentReady || agentConnecting || realtimeUrl.isEmpty) return;
    agentConnecting = true;
    notifyListeners();

    final client = RealtimeClient(
      // Direct Mode drives from here: what the agent returns is plain PCM, fed to
      // the controller exactly as any other PCM16 source would be.
      onAudio: (pcm) async {
        // send(), not yieldAudioData(): that one is Backend Mode's, for audio the
        // server has already driven. Direct Mode drives from here, so the agent's
        // reply goes through send() like any other PCM16 source — `end` stays
        // false, since a turn is many of these and turn_end closes it.
        // Not awaited, and not chained behind the previous chunk: Dart runs this
        // isolate's callbacks one at a time, so the calls reach the platform channel
        // in the order they were made. Waiting for each `send` to return before
        // issuing the next paces the stream to however long the SDK takes to accept
        // a chunk, which comes out as speech that stops between words. The iOS
        // client hands each chunk to the main actor the same way.
        _controller?.send(pcm, end: false);
      },
      onTurnEnd: () async {
        // The empty final send is what tells the SDK the turn is over, so the
        // avatar returns to idle rather than holding the last shape. Queued behind
        // the audio, or it would close a turn whose chunks have not landed yet.
        _controller?.send(Uint8List(0), end: true);
      },
      onInterrupt: () {
        _controller?.interrupt();
      },
      onTranscript: (role, text) {
        if (text.isEmpty) return;
        transcript = [...transcript, (role, text)];
        notifyListeners();
      },
      onError: (message) {
        errorMessage = message;
        onToast?.call(ToastMessage(message));
        notifyListeners();
      },
      onClosed: () {
        agentReady = false;
        micActive = false;
        notifyListeners();
      },
    );

    try {
      // No settings travel with it: the language, the models and the voice are the
      // server's, fixed when it builds the agent session.
      await client.connect(realtimeUrl);
      _realtime = client;
      agentReady = true;
    } catch (e) {
      await client.close();
      errorMessage = e.toString();
      onToast?.call(ToastMessage(e.toString()));
    } finally {
      agentConnecting = false;
      notifyListeners();
    }
  }

  Future<void> startMic() async {
    // The agent drops audio that arrives before it is ready, so opening the mic is
    // what brings it up.
    await _ensureAgent();
    if (!agentReady || micActive) return;

    if (!await _recorder.hasPermission()) {
      errorMessage = 'Microphone permission denied';
      onToast?.call(const ToastMessage('Microphone permission denied'));
      notifyListeners();
      return;
    }

    final stream = await _recorder.startStream(
      const RecordConfig(
        encoder: AudioEncoder.pcm16bits,
        sampleRate: 16000,
        numChannels: 1,
        autoGain: true,
        echoCancel: true,
        noiseSuppress: true,
        // VOICE_COMMUNICATION, not the default MIC: `echoCancel` alone is a software
        // flag, and on Android the hardware canceller only engages on this source.
        // Without it the avatar's own voice comes back in through the microphone,
        // reaches the agent as user speech, and it answers itself.
        androidConfig: AndroidRecordConfig(
          audioSource: AndroidAudioSource.voiceCommunication,
        ),
      ),
    );

    micActive = true;
    notifyListeners();
    _micSubscription = stream.listen((data) => _realtime?.sendMicAudio(data));
  }

  Future<void> stopMic() async {
    if (!micActive) return;
    await _micSubscription?.cancel();
    _micSubscription = null;
    await _recorder.stop();
    micActive = false;
    notifyListeners();
  }

  /// Have the agent speak a typed line — a way to try it without a microphone.
  Future<void> sendText(String text) async {
    final trimmed = text.trim();
    if (trimmed.isEmpty) return;
    await _ensureAgent();
    _realtime?.sendText(trimmed);
  }

  void close() {
    _controller?.close();
  }

  Future<void> closeRealtime() async {
    await stopMic();
    await _realtime?.close();
    _realtime = null;
    agentReady = false;
    agentConnecting = false;
    transcript = [];
  }

  @override
  void dispose() {
    close();
    super.dispose();
  }
}

File: direct-mode/clients/ios/AvatarDemo/ViewModels/AvatarViewModel.swift
import SwiftUI
import Combine
import AVFoundation
import AvatarKit

@MainActor class AvatarViewModel: ObservableObject {
    // What the SDK reports back, one property per public callback. Every one is
    // registered whether or not this demo acts on it: which hooks exist is part of
    // what a reference client is for, and a row that only appears once it has fired
    // is a row nobody knows to expect.
    @Published var connectionState: String = "\(ConnectionState.disconnected)"
    @Published var conversationState: String = "\(ConversationState.idle)"
    /// onFirstRendering — the avatar has actually been drawn.
    @Published var rendered = false
    /// onFrameRateInfo — nil until the monitor has reported once.
    @Published var fps: Int?
    @Published var errorMessage: String?

    @Published var avatar: Avatar?
    @Published var toast: ToastMessage?

    // Conversation.
    @Published var micOn = false
    @Published var agentConnecting = false
    @Published var agentReady = false
    @Published var transcript: [(role: String, text: String)] = []

    private(set) var isConnected = false
    private var avatarController: AvatarController?

    private var realtime: RealtimeClient?
    private var mic: MicrophoneCapture?
    private var realtimeURL = ""

    /// Told once, when the playground opens: where the agent lives. The language, the
    /// models and the voice are the server's `.env` — all three are fixed when it
    /// builds the agent session, so none of them is a client setting.
    func configureRealtime(url: String) {
        realtimeURL = url
    }

    func setAvatarController(_ controller: AvatarController) {
        avatarController = controller
        avatarController?.onConnectionState = { [weak self] state in
            guard let self else { return }
            self.connectionState = "\(state)"
            switch state {
            case .connected:
                self.isConnected = true
            case .disconnected, .failed:
                self.isConnected = false
                // The agent would otherwise hold a model session with nowhere to
                // send its audio.
                self.closeAgent()
            case .connecting:
                break
            @unknown default:
                break
            }
        }
        avatarController?.onConversationState = { [weak self] state in
            self?.conversationState = "\(state)"
        }
        avatarController?.onError = { [weak self] error in
            self?.errorMessage = error.localizedDescription
            self?.toast = ToastMessage(text: error.localizedDescription)
        }
        // Off by default and free while off, so it is switched on here to give the
        // status bar something to report.
        avatarController?.frameRateMonitorEnabled = true
        avatarController?.onFrameRateInfo = { [weak self] info in
            guard let self else { return }
            self.fps = info.fps.isFinite ? Int(info.fps.rounded()) : nil
        }
    }

    func start() { avatarController?.start() }

    func pause() { avatarController?.pause() }
    func resume() { avatarController?.resume() }

    func interrupt() {
        avatarController?.interrupt()
        realtime?.interrupt()
    }

    func close() {
        closeAgent()
        avatarController?.close()
    }

    // MARK: - Conversation

    func toggleMic() async {
        guard isConnected else {
            toast = ToastMessage(text: "Tap Start to connect the avatar first.", kind: .warning)
            return
        }
        if micOn {
            mic?.stop()
            mic = nil
            micOn = false
            return
        }

        guard await requestMicPermission() else {
            toast = ToastMessage(text: "Microphone permission was denied.")
            return
        }
        // The agent is brought up on the first press rather than on entry: it costs
        // a model session, and opening the app should not start one.
        guard await ensureAgent() else { return }

        let capture = MicrophoneCapture(sampleRate: AvatarSDK.configuration.audioFormat.sampleRate)
        do {
            try capture.start { [weak self] pcm in
                self?.realtime?.pushMicAudio(pcm)
            }
            mic = capture
            micOn = true
        } catch {
            toast = ToastMessage(text: error.localizedDescription)
        }
    }

    func say(_ text: String) async {
        let line = text.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !line.isEmpty else { return }
        guard isConnected else {
            toast = ToastMessage(text: "Tap Start to connect the avatar first.", kind: .warning)
            return
        }
        guard await ensureAgent() else { return }
        realtime?.say(line)
    }

    private func ensureAgent() async -> Bool {
        if realtime?.isReady == true { return true }
        guard !realtimeURL.isEmpty else { return false }

        agentConnecting = true
        defer { agentConnecting = false }

        let client = RealtimeClient(callbacks: RealtimeClient.Callbacks(
            onAudio: { [weak self] pcm in
                // Straight to controller.send(). `end`
                // stays false: a turn is many of these, and turn_end closes it.
                Task { @MainActor in self?.avatarController?.send(pcm, end: false) }
            },
            onTurnEnd: { [weak self] in
                // The empty final send is what tells the SDK the turn is over, so the
                // avatar returns to idle rather than holding the last mouth shape.
                Task { @MainActor in self?.avatarController?.send(Data(), end: true) }
            },
            onInterrupt: { [weak self] in
                Task { @MainActor in self?.avatarController?.interrupt() }
            },
            onTranscript: { [weak self] role, text in
                Task { @MainActor in self?.transcript.append((role: role, text: text)) }
            },
            onError: { [weak self] message in
                Task { @MainActor in self?.toast = ToastMessage(text: message) }
            },
            onClosed: { [weak self] in
                Task { @MainActor in
                    self?.agentReady = false
                    self?.micOn = false
                }
            }
        ))

        do {
            // No settings travel with it: the language, the models and the voice are
            // the server's, fixed when it builds the agent session.
            try await client.connect(url: realtimeURL)
            realtime = client
            agentReady = true
            return true
        } catch {
            toast = ToastMessage(text: error.localizedDescription)
            client.close()
            realtime = nil
            return false
        }
    }

    private func requestMicPermission() async -> Bool {
        await withCheckedContinuation { cont in
            // AVAudioApplication is iOS 17+, and this app still supports 16.
            if #available(iOS 17.0, *) {
                AVAudioApplication.requestRecordPermission { granted in
                    cont.resume(returning: granted)
                }
            } else {
                AVAudioSession.sharedInstance().requestRecordPermission { granted in
                    cont.resume(returning: granted)
                }
            }
        }
    }

    private func closeAgent() {
        mic?.stop()
        mic = nil
        micOn = false
        agentReady = false
        realtime?.close()
        realtime = nil
    }
}
